### ISSUE TITLE
Android Platform Support

### ISSUE THREAD

**Issue by daylightwarbler** [BurntSushi/jiff#140]
Getting the current time works as expected, but using the system Time Zone Database and obtaining the system time zone both require additional crates and lower-level Jiff methods. If Android was supported out of the box, then that would also improve ergonomics for cross-platform Android/iOS projects, since Jiff's API could be used the same way on both targets.

On Android, the Time Zone Database is in an Android-specific format. The [`android-tzdata`](https://crates.io/crates/android-tzdata) crate provides TZif data for a time zone name.

```rust
use jiff::{civil::date, tz::TimeZone};

let tz_name = "America/New_York";
let tz_data = android_tzdata::find_tz_data(tz_name)?;
let tz = TimeZone::tzif(tz_name, &tz_data)?;
let zdt = date(2023, 12, 31).at(18, 30, 0, 0).to_zoned(tz)?;
assert_eq!(zdt.to_string(), "2023-12-31T18:30:00-05:00[America/New_York]");
```

The system time zone can be obtained with the [`iana-time-zone`](https://crates.io/crates/iana-time-zone) crate.

```rust
use jiff::{tz::TimeZone, Timestamp};

let tz_name = iana_time_zone::get_timezone()?;
let tz_data = android_tzdata::find_tz_data(&tz_name)?;
let tz = TimeZone::tzif(&tz_name, &tz_data)?;
let zdt_now = Timestamp::now().to_zoned(tz);
```

**Comment by BurntSushi** [comment]
Yeah I do want to support this automatically. But I wanted a concrete thing I could test somehow first. Is there an example Rust program along with perhaps a CI setup that demonstrates the full end to end need here?

Basically, I haven't written Rust programs on Android before. I do have a good idea of what needs to be done based on prior art research, but I don't know exactly how I should go about iterating and testing. If someone can help me with that part, then I can do the rest.

**Comment by daylightwarbler** [comment]
Ok, here’s an example Android app build with Jiff and a Github actions CI workflow. The app runs some [Rust code](https://github.com/daylightwarbler/android-rust-example/blob/main/rust/src/lib.rs) at startup and then shows some text. The test installs and starts the app in a running emulator and checks if the app crashed or not. If an assert fails then that causes a crash and test failure.

https://github.com/daylightwarbler/android-rust-example

**Comment by BurntSushi** [comment]
Thank you! I managed to get that to work after a lot of faffing about. Holy shit what a complex nightmare. But I think that's enough for me to iterate. Thanks again.

**Comment by BurntSushi** [comment]
Oh, is there an easy way to make `BAZEL` build a copy of Jiff from my local disk instead of taking it from the crate index? Specifically, this part:

https://github.com/daylightwarbler/android-rust-example/blob/5327693d1afba0213b5ce9b7ef029120feb6c579/rust/BUILD#L4-L15

**Comment by BurntSushi** [comment]
OK, I managed to get this working. It should be available automatically in `jiff 0.1.22` on crates.io. Please see the [platform docs for Android](https://docs.rs/jiff/latest/jiff/_documentation/platform/index.html#android) for more specific details.

**Comment by dead10ck** [comment]
Looks like this does not compile on-device from a Termux shell. There are 5 instances of this error:

```
error[E0308]: mismatched types
   --> /data/data/com.termux/files/home/.cargo/registry/src/index.crates.io
-6f17d22bba15001f/jiff-0.1.22/src/tz/system/android.rs:117:38
    |
117 | ...= unsafe { dlopen(cstr("libc.so\0").as_ptr(), 0) };
    |               ------ ^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `*const i8`,
 found `*const u8`
    |               |
    |               arguments to this function are incorrect
    |
    = note: expected raw pointer `*const i8`
               found raw pointer `*const u8`
note: function defined here
   --> /data/data/com.termux/files/home/.cargo/registry/src/index.crates.io
-6f17d22bba15001f/jiff-0.1.22/src/tz/system/android.rs:293:8
    |
293 |     fn dlopen(filename: *const i8, flag: i32) -> *mut c_void;
    |        ^^^^^^
```

**Comment by BurntSushi** [comment]
Wat. Can you open a new issue and give steps for a reproduction please?

**Comment by BurntSushi** [comment]
Also, can you include the target triple?

**Comment by BurntSushi** [comment]
I created a new issue for that in #200. But would still love a repro that I can follow along with (speaking as someone who has never used termux before) and a target triple so that I can see about adding it to CI.
